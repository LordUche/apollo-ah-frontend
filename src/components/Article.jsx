import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import ArticleBody from '../views/ArticleBody';
import { fetchArticle } from '../actions/articleAction';
import { getArticle } from '../selectors/articlesSelector';

/*
 * @todo - Import comments component here
 */

/**
 *
 *
 * @class Article
 * @extends {React.Component}
 */
class Article extends React.Component {
  /**
   *
   * @static - Validating proptypes
   * @memberof Article
   */
  static propTypes = {
    match: PropTypes.object.isRequired,
    article: PropTypes.object.isRequired,
    loadArticle: PropTypes.func.isRequired,
  }

  /**
   * Creates an instance of Article.
   * @param {*} props - Props passed down from the BrowserRouter
   * @memberof Article
   */
  constructor(props) {
    super(props);
    this.state = {
      bookmarked: false,
    };
  }

  /**
   *
   * @todo - Move the fetch article functionality to redux store
   * @memberof Article
   */
  componentDidMount = () => {
    const { match, loadArticle } = this.props;
    const { slug } = match.params;
    loadArticle(slug);
  }

  /**
   * @todo - employ javascript debounce to avoid calling API on repeated clicks
   * @method bookmarkArticle - The bookmark article functionality
   * @memberof Article
   */
  bookmarkArticle = () => {
    const { bookmarked } = this.state;
    this.setState({ bookmarked: !bookmarked });
  };

  /**
   *
   *
   * @returns {JSX} The JSX representation of the Article component
   * @memberof Article
   */
  render() {
    const { bookmarked } = this.state;
    const { article } = this.props;
    return (
      <div>
        <main className="main-body">
          <section className="min-vh-100">
            {article.image && (
              <div
                className="pg-empty-placeholder single-hero"
                style={{ backgroundImage: `url(${article.image})` }}
              />
            )}
            <div className="container">
              <div className="single-container">
                <ArticleBody
                  article={article}
                  bookmarkArticle={this.bookmarkArticle}
                  bookmarked={bookmarked}
                  applaud={this.applaud}
                />
                {/* Insert omment component here */}
              </div>
            </div>
            <div className="single-suggested-grp">
              <div className="container">
                <h4>Recommended for you</h4>
                <div className="row pg-empty-placeholder" />
              </div>
            </div>
          </section>
        </main>
      </div>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  article: getArticle,
});

const mapDispatchToProps = {
  loadArticle: slug => fetchArticle(slug),
};

export default connect(mapStateToProps, mapDispatchToProps)(Article);
