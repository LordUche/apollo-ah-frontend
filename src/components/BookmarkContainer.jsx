/* eslint-disable react/jsx-indent */
/* eslint-disable react/jsx-closing-tag-location */
/* eslint-disable react/no-array-index-key */
/* eslint-disable react/jsx-wrap-multilines */
import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import BookmarkItem from '../views/BookmarkItem';
import { getBookmarkedArticles } from '../actions/getBookmark';
import * as selectors from '../selectors/getBookmarkSelector';

/**
 * @description - returns all articles bookmarked by a user
 * @return {JSX}
 */
function BookmarkContainer(props) {
  const [viewType, setViewType] = useState('row view-row');

  const { articles } = props;

  const getArticlesHandler = () => {
    props.getBookmarkedArticles();
  };

  const changeViewToHorizontal = () => {
    setViewType('row view-row list-view');
  };

  const changeViewToVertical = () => {
    setViewType('row view-row');
  };

  useEffect(() => (
    getArticlesHandler()
  ), []);

  return (
    <>
      <header className="page-header" data-pg-collapsed>
        <h1>Bookmarks </h1>
        <div className="btn-group btn-group-toggle float-right view-changer" data-toggle="buttons">
          <label className="btn active" onClick={changeViewToVertical}>
            <input type="radio" name="options" autoComplete="off" checked value="" />
            <i className="fas fa-th" />
          </label>
          <label className="btn" onClick={changeViewToHorizontal}>
            <input type="radio" name="options" autoComplete="off" value="list-view" />
            <i className="fas fa-list" />
          </label>
        </div>
      </header>
      {
        articles.articles.length !== 0
          ? (
            <>
              <div className={viewType}>
                {
                  // eslint-disable-next-line arrow-body-style
                  articles.articles.map((article, indx) => {
                    return (
                      <BookmarkItem
                        key={indx}
                        category={article.Article.articleCategory.category}
                        image={article.Article.image}
                        articleTitle={article.Article.title}
                        articleDescription={article.Article.description}
                        authorsName={article.Article.User.Profile.firstname}
                        authorsImage={article.Article.User.Profile.image}
                        readTime={article.Article.readTime}
                      />
                    );
                  })
                }
              </div>
              {/** pagination support to be implemented later */}
              <div>Pagination is here</div>
            </>
          ) : <div>
            <i className="fas fa-book-open" style={{ fontSize: '70px', color: '#66008c' }} />
            <p style={{ paddingTop: '30px' }}>*It looks like you have not bookmarked any article yet</p>
          </div>
      }
    </>
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getBookmarkLoading,
    articles: selectors.getBookmarkArticles,
  }
);

export default connect(mapStateToProps, { getBookmarkedArticles })(BookmarkContainer);
