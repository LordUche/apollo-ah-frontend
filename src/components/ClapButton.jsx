import React, { Component } from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { debounce } from 'lodash';
import clapIcon from '../images/clap.svg';
import { getArticle } from '../selectors/articlesSelector';
import { getUserClaps } from '../selectors/clapsSelector';
import { fetchUserClaps } from '../actions/clapsAction';
import { clapArticleRequest } from '../actions/articleAction';

/**
 * A class representing clap button component
 *
 * @class ClapButton
 * @extends {Component}
 */
export class ClapButton extends Component {
  static propTypes = {
    article: PropType.object.isRequired,
    clapArticle: PropType.func.isRequired,
    loadUserClaps: PropType.func.isRequired,
    claps: PropType.number.isRequired,
  };

  /**
   * Handles clap request (debounced)
   *
   * @returns {void}
   * @memberof ClapButton
   */
  makeClapRequest = debounce(() => {
    const { clapArticle, article: { slug } } = this.props;
    const { claps } = this.state;
    this.setState({ claps: 0 });
    return clapArticle({ slug, claps });
  }, 1000);

  constructor(props) {
    super(props);
    this.state = {
      clapsLeft: 100 - props.claps,
      articleClaps: props.article.claps,
      claps: 0, // record clap button clicks
    };
  }

  componentDidMount = () => {
    const { article: { slug }, loadUserClaps } = this.props;
    loadUserClaps(slug);
  }

  /**
   * Handles button click
   *
   * @returns {void}
   * @memberof ClapButton
   */
  handleClick = () => {
    const { articleClaps, clapsLeft, claps } = this.state;
    if (clapsLeft >= 1) {
      this.setState({
        clapsLeft: clapsLeft - 1,
        articleClaps: articleClaps + 1,
        claps: claps + 1,
      });
      this.makeClapRequest();
    }
  }

  /**
   * Handles button click
   *
   * @returns {JSX.Element} DOM element
   * @memberof ClapButton
   */
  render() {
    const { articleClaps } = this.state;

    return (
      <span className="clap-icon">
        <button type="button" className="btn" onClick={this.handleClick}>
          <img src={clapIcon} alt="" />
        </button>
        {articleClaps}
        {' '}
        claps
      </span>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  article: getArticle,
  claps: getUserClaps,
});

const mapDispatchToProps = {
  clapArticle: claps => clapArticleRequest(claps),
  loadUserClaps: payload => fetchUserClaps(payload),
};

export default connect(mapStateToProps, mapDispatchToProps)(ClapButton);
