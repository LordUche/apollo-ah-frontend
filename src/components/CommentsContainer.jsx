/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';

import Comments from '../views/Comments';

/**
 * Container component for the Article comments view
 * @export
 * @class CommentsContainer
 * @extends {Component}
 */
export default class CommentsContainer extends Component {
  render() {
    return (
      <Comments />
    );
  }
}
