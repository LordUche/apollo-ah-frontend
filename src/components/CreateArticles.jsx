/* eslint-disable no-unused-vars */
/* eslint-disable arrow-parens */
/* eslint-disable react/jsx-no-bind */
/* eslint-disable jsx-a11y/no-redundant-roles */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React, { useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import * as selectors from '../selectors/createArticleSelector';
import createArticleAction from '../actions/createArticles';
import CreateArticle from '../views/CreateArticle';
import CloudinaryWidget from './common/CloudinaryWidget';


/**
 * Form for creating a new article
 * @param {object} props Component props
 * @param {function} props.createArticleAction action creator function
 */
function CreateArticleContainer(props) {
  const [formData, setFormData] = useState({
    title: null,
    description: null,
    body: null,
    categoryId: null,
    tagList: [],
    image: null,
  });

  /** passing global state loading as props into variables */
  let { message, isLoading, token } = props;

  let widget;

  const updateInputHandler = (event) => {
    setFormData({ ...formData, [event.target.name]: event.target.value });
  };

  const updateBodyHandler = (value) => {
    setFormData({ ...formData, body: value });
  };

  const updateTagHandler = (tags) => {
    const listOfTags = [];
    tags.forEach(tag => {
      listOfTags.push(tag.text);
      setFormData({ ...formData, tagList: listOfTags });
    });
  };

  const submitHandler = (event) => {
    event.preventDefault();
     console.log('--->', formData);
     const requestData = {
       formData,
       token,
     };
    //props.createArticleAction(formData);
    props.createArticleAction(requestData);
  };

  const cloudinaryUpdate = (imageUrl) => {
    setFormData({ ...formData, image: imageUrl });
  };

  const imageUploadHandler = () => {
    // console.log('yo mama');
    // const { showLoader } = this.props;
    // showLoader(true);
    if (widget === undefined) {
      widget = new CloudinaryWidget(
        cloudinaryUpdate, (error) => {
        return (error);
      }, false);
    }
    widget.open();
  };

  console.log('--->', isLoading);
  console.log('==->', message);

  return (
    <CreateArticle
      updateInputHandler={updateInputHandler}
      updateBodyHandler={updateBodyHandler}
      updateTagHandler={updateTagHandler}
      imageUploadHandler={imageUploadHandler}
      submitHandler={submitHandler}
      valueText={formData.body}
    />
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getCreateArticleLoadingState,
    message: selectors.getCreateArticleMessageState,
    token: selectors.getUserToken,
  }
);

export default connect(mapStateToProps, { createArticleAction })(CreateArticleContainer);
