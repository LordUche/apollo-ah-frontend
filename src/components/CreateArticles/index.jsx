/* eslint-disable jsx-a11y/no-redundant-roles */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React from 'react';

function CreateArticle(props) {
  return (
    <section className="wrapper min-vh-100">
      <div className="container">
        <header className="page-header" data-pg-collapsed>
          <h1> New Article</h1>
        </header>
        <div className="card mb-5 mt-n5">
          <form role="form" className="form p-sm-4 p-3 form-borderless">
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="title" className="font-weight-bold">Title</label>
              <input type="text" className="form-control text-capitalize" id="title" name="title" required />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="description" className="font-weight-bold">Short Description</label>
              <textarea className="form-control textarea-sm" id="description" name="description" />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="description" className="font-weight-bold">Details</label>
              <textarea className="form-control" id="body" name="body" />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label className="font-weight-bold" htmlFor="category">Category</label>
              <select className="custom-select" name="category" id="category">
                <option selected value="-1">-- Choose one --</option>
                <option value="3">Business</option>
                <option value="7">Entertainment</option>
                <option value="4">Fashion</option>
                <option value="6">Food</option>
                <option value="5">Health</option>
                <option value="8">History</option>
                <option value="1">Technology</option>
                <option value="2">Sports</option>
                <option value="9">Others</option>
              </select>
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label className="font-weight-bold" htmlFor="tags">Tags (optional)</label>
              <input type="text" className="form-control" id="tags" name="tags" data-role="tagsinput" />
            </div>
            <div className="form-group mb-4">
              <div className="custom-file">
                <input type="file" className="custom-file-input" id="custom-file1" />
                <label className="custom-file-label" htmlFor="custom-file1">Upload cover image</label>
              </div>
            </div>
            <button type="submit" className="btn btn-brand btn-plain mt-3">Publish article</button>
          </form>
        </div>
      </div>
    </section>
  );
}

export default CreateArticle;
