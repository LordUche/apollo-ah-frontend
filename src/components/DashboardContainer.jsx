import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { getDashboard, getDashboardSummary } from '../selectors/dashboardSelector';
import { getProfile } from '../selectors/profileSelector';
import Dashboard from '../views/Dashboard';
import ProfileHeaderContainer from './ProfileHeaderContainer';
import { getDashboardArticles } from '../actions/dashboardAction';
import CommentLoader from '../views/CommentLoader';

/**
 * Container component for the Dashboard view
 * @export
 * @class DashboardContainer
 * @extends {Component}
 */
class DashboardContainer extends Component {
  async componentDidMount() {
    const { profile: { id } } = this.props;
    const { getDashboardArticles: fetchDashboardArticles } = this.props;
    await fetchDashboardArticles(id);
  }

  render() {
    const { dashboard, summary } = this.props;
    return (
      <>
        <ProfileHeaderContainer activePage="DASHBOARD" />
        {
          dashboard.loading
            ? <CommentLoader />
            : <Dashboard summary={summary} dashboard={dashboard} />
        }
      </>
    );
  }
}

DashboardContainer.propTypes = {
  dashboard: PropTypes.object.isRequired,
  summary: PropTypes.object.isRequired,
  profile: PropTypes.object.isRequired,
  getDashboardArticles: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector(
  {
    dashboard: getDashboard,
    summary: getDashboardSummary,
    profile: getProfile,
  }
);
const mapDispatchToProps = dispatch => ({
  getDashboardArticles: payload => dispatch(getDashboardArticles(payload)),
});
export default connect(mapStateToProps, mapDispatchToProps)(DashboardContainer);
