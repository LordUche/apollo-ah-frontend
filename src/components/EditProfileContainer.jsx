import React, {Component} from 'react';
import { connect } from 'react-redux';

import EditProfile from '../views/EditProfile';
import ProfileHeaderContainer from './ProfileHeaderContainer';
import Wrapper from '../views/reuse/Wrapper';
import {updateUserProfile} from '../actions/updateProfileAction';


/**
 * Container component for the EditProfile view
 * @export
 * @class EditProfileContainer
 * @extends {Component}
 */
  class EditProfileContainer extends Component{
     /**
     *Creates an instance of EditProfileContainer.
     * @param {object} props The props object
     * @memberof EditProfileContainer
     */
    constructor(props){
        super(props);
        this.handleUpdateProfile = this.handleUpdateProfile.bind(this);
    }

    handleUpdateProfile(event){
      event.preventDefault();
      const data = new FormData(event.target);
      const {updateProfile} = this.props;
      updateProfile(data);
    }
    
     /**
     * The render method
     * @returns {array} The resulting JSX object
     * @memberof EditProfileContainer
     */
    render(){
      const {isLoading, profile, errorData} = this.props;
        return (
          <Wrapper>
            <ProfileHeaderContainer activePage="PROFILE" />
            <EditProfile 
            handleUpdateProfile={this.handleUpdateProfile} 
            isLoading={isLoading} 
            profile={profile}
            errorData={errorData}
            />
          </Wrapper>
        );
    }
}

const mapStateToProps = state => ({
  isLoading: state.loading,
  profile: state.user,
  errorData: state.user.error,
});

const mapDispatchToProps = dispatch => ({
  updateProfile: userData => dispatch(updateUserProfile(userData)),
});
export default connect(mapStateToProps, mapDispatchToProps)(EditProfileContainer);
