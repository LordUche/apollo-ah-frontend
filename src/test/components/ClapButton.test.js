/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ClapButton } from '../../components/ClapButton';

/**
 * Wrapper for enzyme shallow component
 *
 * @returns {object} The mocked props and component
 */
function setup() {
  const props = {
    article: {},
    clapArticle: jest.fn(),
    loadUserClaps: jest.fn(),
    claps: 0,
  };

  const enzymeWrapper = shallow(<ClapButton {...props} />);
  return { props, enzymeWrapper };
}

describe('<ClapButton />', () => {
  it('renders without crashing given the required props', () => {
    const { enzymeWrapper } = setup();
    expect(toJson(enzymeWrapper)).toMatchSnapshot();
  });
});
