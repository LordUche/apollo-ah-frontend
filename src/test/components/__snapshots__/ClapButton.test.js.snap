// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClapButton /> renders without crashing given the required props 1`] = `
<span
  className="clap-icon"
>
  <button
    className="btn"
    onClick={[Function]}
    type="button"
  >
    <img
      alt=""
      src={Object {}}
    />
  </button>
   
  claps
</span>
`;
